include/hc_printf.hpp

```diff
@@ -111,56 +102,93 @@ static inline void copy_n(char* dest, const char* src, unsigned int len) [[hc,cp
   }
   copy_n(&string_buffer[sb_offset], string, str_len + 1);
-  queue[offset].set(&string_buffer[sb_offset]);
+  queue[poffset].set(&string_buffer[sb_offset]);
+  soffset += str_len + 1;
   return PRINTF_SUCCESS;
 }
 
 template <typename T>
-static inline PrintfError set_batch(PrintfPacket* queue, int offset, const T t) [[hc,cpu]] {
+static inline PrintfError set_batch(PrintfPacket* queue, int poffset, unsigned int& soffset, const T t) [[hc,cpu]] {
   PrintfError err = PRINTF_SUCCESS;
-  queue[offset].set(t);
-  if (queue[offset].type == PRINTF_CHAR_PTR || queue[offset].type == PRINTF_CONST_CHAR_PTR){
-    err = process_str_batch(queue, offset, (char*)t);
+  queue[poffset].set(t);
+  if (queue[poffset].type == PRINTF_CHAR_PTR || queue[poffset].type == PRINTF_CONST_CHAR_PTR){
+    err = process_str_batch(queue, poffset, soffset, (char*)t);
   }
   return err;
 }
 template <typename T, typename... Rest>
-static inline PrintfError set_batch(PrintfPacket* queue, int offset, const T t, Rest... rest) [[hc,cpu]] {
+static inline PrintfError set_batch(PrintfPacket* queue, int poffset, unsigned int& soffset, const T t, Rest... rest) [[hc,cpu]] {
   PrintfError err = PRINTF_SUCCESS;
-  queue[offset].set(t);
-  if (queue[offset].type == PRINTF_CHAR_PTR || queue[offset].type == PRINTF_CONST_CHAR_PTR){
-    if ((err = process_str_batch(queue, offset, (char*)t)) != PRINTF_SUCCESS)
+  queue[poffset].set(t);
+  if (queue[poffset].type == PRINTF_CHAR_PTR || queue[poffset].type == PRINTF_CONST_CHAR_PTR){
+    if ((err = process_str_batch(queue, poffset, soffset, (char*)t)) != PRINTF_SUCCESS)
       return err;
   }
-  return set_batch(queue, offset + 1, rest...);
+  return set_batch(queue, poffset + 1, soffset, rest...);
 }
 
 template <typename... All>
 static inline PrintfError printf(PrintfPacket* queue, All... all) [[hc,cpu]] {
-  unsigned int count = 0;
-  countArg(count, all...);
+  unsigned int count_arg = 0;
+  unsigned int count_char = 0;
+  countArg(count_arg, count_char, all...);
 
   PrintfError error = PRINTF_SUCCESS;
+  PrintfPacketData old_off, try_off;
+  unsigned long long old_long;
 
-  if (!queue || count + 1 + queue[1].data.ui > queue[0].data.ui) {
+  if (!queue || count_arg + 1 + queue[3].data.ai[0] > queue[0].data.ui) {
     error = PRINTF_BUFFER_OVERFLOW;
-  } else {
-
-    unsigned int offset = queue[1].data.ai.fetch_add(count + 1);
-    if (offset + count + 1 < queue[0].data.ui) {
-      if (set_batch(queue, offset, count, all...) != PRINTF_SUCCESS)
-        error = PRINTF_STRING_BUFFER_OVERFLOW;
+  }
+  else if (!queue[1].data.ptr || count_char + queue[3].data.ai[1] > queue[2].data.ui){
+    error = PRINTF_STRING_BUFFER_OVERFLOW;
+  }
+  else {
+    do {
+	  old_off.al = queue[3].data.al.load();
+	  old_long = old_off.al.load();
+	  try_off.ai[0] = old_off.ai[0] + count_arg + 1;
+	  try_off.ai[1] = old_off.ai[1] + count_char;
+    } while(!(queue[3].data.al.compare_exchange_weak(old_long, (unsigned long long)try_off.al.load())));
```

Comment:
old_off and try_off are local to the thread and therefore they don't need to be defined with an atomic type